key_management/src/keystore.rs

```diff
@@ -412,59 +410,77 @@ impl KeyStore {
 impl EncryptedKeyStore {
     fn derive_key(
         passphrase: &str,
-        prev_salt: Option<Vec<u8>>,
-    ) -> Result<(pwhash::Salt, Arc<secretbox::Key>), EncryptedKeyStoreError> {
+        prev_salt: Option<[u8; RECOMMENDED_SALT_LEN]>,
+    ) -> anyhow::Result<([u8; RECOMMENDED_SALT_LEN], Vec<u8>)> {
         let salt = match prev_salt {
-            Some(prev_salt) => match pwhash::Salt::from_slice(&prev_salt) {
-                Some(salt) => salt,
-                None => return Err(EncryptedKeyStoreError::ConfigurationError),
-            },
-            None => pwhash::gen_salt(),
+            Some(prev_salt) => prev_salt,
+            None => {
+                let mut salt = [0; RECOMMENDED_SALT_LEN];
+                OsRng.fill_bytes(&mut salt);
+                salt
+            }
         };
-        let mut key = secretbox::Key([0; secretbox::KEYBYTES]);
-
-        let secretbox::Key(ref mut kb) = key;
-        pwhash::derive_key(
-            kb,
-            passphrase.as_bytes(),
-            &salt,
-            pwhash::OPSLIMIT_INTERACTIVE,
-            pwhash::MEMLIMIT_INTERACTIVE,
-        )
-        .unwrap();
-
-        Ok((salt, Arc::new(key)))
-    }
-
-    fn encrypt(encryption_key: Arc<secretbox::Key>, msg: &[u8]) -> Vec<u8> {
-        let nonce = secretbox::gen_nonce();
 
-        let mut ciphertext = secretbox::seal(msg, &nonce, &encryption_key);
-        ciphertext.append(&mut nonce.as_ref().to_vec());
-        ciphertext
+        let mut param_builder = ParamsBuilder::new();
+        // #define crypto_pwhash_argon2id_MEMLIMIT_INTERACTIVE 67108864U
+        param_builder
+            .m_cost(67108864 / 1024)
+            .map_err(map_err_to_anyhow)?;
+        // #define crypto_pwhash_argon2id_OPSLIMIT_INTERACTIVE 2U
+        param_builder.t_cost(2).map_err(map_err_to_anyhow)?;
+        // https://docs.rs/sodiumoxide/latest/sodiumoxide/crypto/secretbox/xsalsa20poly1305/constant.KEYBYTES.html
+        // KEYBYTES = 0x20
+        // param_builder.output_len(32)?;
```

Comment:
Do we need these leftover comments? I mean, it's helpful to know where did `67108864U` come from; perhaps a link to the C/C++ implementation where it's defined would be more beneficial for the reader.

Reply:
links added